std-bits/base/src/main/java/org/enso/base/parser/NumberWithSeparators.java

```diff
@@ -0,0 +1,393 @@
+    }
+
+    // Strip out the separators.
+    int origEndIdx = endIdx;
+    if (thousands != Constants.NONE) {
+      value = Separators.strip(value, idx, endIdx, thousands, decimal);
+      if (value == null) {
+        return new NumberParseFailure("Invalid number.");
+      }
+      idx = 0;
+      endIdx = value.length();
+    }
+
+    try {
+      long number = Long.parseLong(value, idx, endIdx, 10);
+      return new NumberParseResultWithIndex(origEndIdx, new NumberParseLong(number, ""));
+    } catch (NumberFormatException e) {
+      return new NumberParseFailure("Invalid number.");
+    }
+  }
+
+  /** Parse an unknown format with no separators. */
+  private NumberParseResult parseUnknownIntegerNone(CharSequence value, int idx, int endIdx) {
+    assert thousands == Constants.UNKNOWN;
+
+    // We haven't encountered any separators. So parse the number as a long.
+    try {
+      long number = Long.parseLong(value, idx, endIdx, 10);
+      var result = new NumberParseResultWithIndex(endIdx, new NumberParseLong(number, ""));
+
+      // If greater than or equal 1000, then we know no thousand separators.
+      if (number >= 1000) {
+        var format =
+            switch (decimal) {
+              case '.' -> NO_DOT;
+              case ',' -> NO_COMMA;
+              default -> NO_UNKNOWN;
+            };
+
+        if (this != format) {
+          return new NumberParseResultWithFormat(format, result);
+        }
+      }
+
+      return result;
+    } catch (NumberFormatException e) {
+      return new NumberParseFailure("Invalid number.");
+    }
+  }
+
+  /** Parse an unknown Integer format. */
+  private NumberParseResult parseUnknownInteger(
+      CharSequence value, int idx, int endIdx, char separator, int separatorCount) {
+    assert thousands == Constants.UNKNOWN;
+
+    if (separator == Constants.NONE) {
+      // Didn't encounter any separators so use simpler logic.
+      return parseUnknownIntegerNone(value, idx, endIdx);
+    }
+
+    // Find the correct format
+    var format =
+        switch (separator) {
+          case '.' -> DOT_COMMA;
+          case ',' -> separatorCount == 1 ? COMMA_UNKNOWN : COMMA_DOT;
+          case ' ' -> (decimal == Constants.UNKNOWN
+              ? SPACE_UNKNOWN
+              : (decimal == '.' ? SPACE_DOT : SPACE_COMMA));
+          case '\'' -> (decimal == Constants.UNKNOWN
+              ? SWISS_UNKNOWN
+              : (decimal == '.' ? SWISS_DOT : SWISS_COMMA));
+          default -> null;
+        };
+    if (format == null) {
+      return new NumberParseFailure("No matching number format.");
+    }
+
+    var result = format.parseFixedInteger(value, idx, endIdx, separator);
+    return (result instanceof NumberParseFailure)
+        ? result
+        : new NumberParseResultWithFormat(format, result);
+  }
+
+  /** Given a known double format, parse the sequence. */
+  private NumberParseResult parseFixed(
```

Comment:
nitpick stylistic suggestion:

```suggestion
  private NumberParseResult parseFixedDecimal(
```

will make it more explicit which methods are for integers which for decima.